### ISSUE TITLE
Error message when wrong call to forward of RNN (nn.GRU and also nn.LSTM)

### ISSUE THREAD

**Issue by Diego999** [pytorch/pytorch#5211]
Hi everyone,

I'm implementing a model with PyTorch and realised when I call my GRU forward method, I raise an exception (due to bad shape) but there is an error in the message given due to undefined fn:

Traceback (most recent call last):
...
    output, hidden = self.document_rnn(sentence_hidden_embeddings, hidden)
  File "/usr/local/lib/python3.6/site-packages/torch/nn/modules/module.py", line 357, in __call__
    result = self.forward(*input, *_*kwargs)
  File "/usr/local/lib/python3.6/site-packages/torch/nn/modules/rnn.py", line 190, in forward
    self.check_forward_args(input, hx, batch_sizes)
  **File "/usr/local/lib/python3.6/site-packages/torch/nn/modules/rnn.py", line 141, in check_forward_args
    fn.input_size, input.size(-1)))
NameError: name 'fn' is not defined**

In the original code source:

`
def check_forward_args(self, input, hidden, batch_sizes):

        is_input_packed = batch_sizes is not None
        expected_input_dim = 2 if is_input_packed else 3

        if input.dim() != expected_input_dim:
            raise RuntimeError(
                'input must have {} dimensions, got {}'.format(
                    expected_input_dim, input.dim()))

        if self.input_size != input.size(-1):
            raise RuntimeError(
                'input.size(-1) must be equal to input_size. Expected {}, got {}'.format(
                    fn.input_size, input.size(-1)))
        if is_input_packed:
            mini_batch = batch_sizes[0]
        else:
            mini_batch = input.size(0) if self.batch_first else input.size(1)

        num_directions = 2 if self.bidirectional else 1
        expected_hidden_size = (self.num_layers * num_directions,
                                mini_batch, self.hidden_size)

        def check_hidden_size(hx, expected_hidden_size, msg='Expected hidden size {}, got {}'):
            if tuple(hx.size()) != expected_hidden_size:
                raise RuntimeError(msg.format(expected_hidden_size, tuple(hx.size())))

        if self.mode == 'LSTM':
            check_hidden_size(hidden[0], expected_hidden_size,
                              'Expected hidden[0] size {}, got {}')
            check_hidden_size(hidden[1], expected_hidden_size,
                              'Expected hidden[1] size {}, got {}')
        else:
            check_hidden_size(hidden, expected_hidden_size)`
`
fn is undefined. It should be self.input_size.

- OS: Mac
- PyTorch version: 0.3.0
- How you installed PyTorch (conda, pip, source): pip
- Python version: 3.6

**Comment by zou3519** [comment]
Thanks for the report! This has been fixed on master ([see file here](https://github.com/pytorch/pytorch/blob/2b2d56d8460d335daf5aa79774442a111d424f90/torch/nn/modules/rnn.py#L118))

**Comment by maartjeth** [comment]
I just had to freshly install PyTorch and I get exactly this error now. Same details, but on Ubuntu 17.10 and while using nn.LSTM.

**Comment by zou3519** [comment]
@maartjeth what version of pytorch are you using? `python -c "import torch; print(torch.__version__)"`. This issue should have been fixed on master.

**Comment by maartjeth** [comment]
@zou3519 0.3.1

**Comment by the-new-sky** [comment]
I have the same issue on PyTorch 0.3.1 using an LSTM. Tried to reinstall it from PyPI but no change, should I compile it from source to get it work ?

**Comment by zou3519** [comment]
Could one of you create a minimal script that reproduces the error?

**Comment by the-new-sky** [comment]
After a few attempts, it turns out that, at least in my case, the error occurs when the dimension 2 of the input is not correct.
```python
import torch
from torch import nn
from torch.autograd import Variable

model = nn.LSTM(5, 1)

input = Variable(torch.rand(1, 1, 3))  # input = Variable(torch.rand(1, 1, 5)) to get it work :)
h0 = Variable(torch.rand(1, 1, 1))
c0 = Variable(torch.rand(1, 1, 1))

print(model(input, (h0, c0)))
```

**Comment by zou3519** [comment]
@the-new-sky that particular bug (`fn` not being defined) was fixed on master, so yes, building from source will make that throw a better error message :)

**Comment by erikreppel** [comment]
Are there any updates on this? Makes it quite hard to work with LSTMs and GRUs with pytorch right now

**Comment by swathirajanna** [comment]
This is still an issue in torch version 0.3.1.post2

**Comment by zou3519** [comment]
Yes, it will be fixed in the 0.4 release.

**Comment by jrzaurin** [comment]
I just downgraded to 0.3.0 and all fine

`conda install pytorch=0.3.0 torchvision cuda91 -c soumith`

please,  let us know when the release happens :)
